File: modules_lib/bleextop/libraries/applicationbi.py
# -*- coding: utf-8 -*-
import tree
from modules_lib.bleextop.models import applicationdao
import datetime


class ApplicationBI:

    def __init__(self):
        self.appdao = applicationdao.ApplicationDAO()

    def getApplications(self, username):
        apps, confs = self.appdao.getApplications(username)
        #print apps
        thisTree = self.buildTree(apps, confs, "menu")
        return thisTree.getRoot()

    def getTree(self):
        apps, confs = self.appdao.getAllApps()
        #print apps, confs
        thisTree = self.buildTree(apps, confs, "children")
        return thisTree.getRoot()

    def saveApp(self, data):
        data["date_updated"] = datetime.datetime.now().strftime("%Y-%m-%d %H:%M:%S")
        data["date_created"] = datetime.datetime.now().strftime("%Y-%m-%d %H:%M:%S")
        self.appdao.saveApp(data)
        appId = self.appdao.getInsertId()
        self.appdao.addPermissions(appId)

        return [{"success": True},
                {"application_k": id},
                {"message": "Application successfully saved"}]

    def updateApp(self, data):
        data["date_updated"] = datetime.datetime.now().strftime("%Y-%m-%d %H:%M:%S")
        success = self.appdao.updateApp(data)
        if success:
            return [{"success": True}, {"application_k": data["application_k"]}, {"message": "Application successfully updated"}]
        else:
            return [{"success": False}, {"message": "There was an error updating this application."}]

    def removeApp(self, application_k):
        self.appdao.removeApp(application_k)

    def moveApp(self, data):
        success = self.moveApp(data)
        if success:
            return [{"success": True}, {"message": "Application successfully moved"}]
        else:
            return [{"success": False}, {"message": "There was an error moving this application."}]

    def buildTree(self, apps, confs, text):
        temp = []
        count = 0
        #print apps
        for app in apps:
            iconCls = ""
            if app["configurations"]:
                conf = confs[count]
                if conf["iconCls"]:
                    iconCls = conf["iconCls"]
            temp.append({"text": app["name"],
                    "name": app["name"],
                    "application_k": app["application_k"],
                    "application_parent_k": app["application_parent_k"],
                    "klass": app["klass"],
                    "description": app["description"],
                    "configurations": app["configurations"],
                    "active": app["active"],
                    "iconCls": iconCls})
            count += 1

        # Creating the Tree
        thisTree = tree.Tree()
        thisTree.setChildProperty(text)
        thisTree.setIdProperty("application_k")
        for app in temp:
            #print app
            thisTree.addChild(app, app["application_parent_k"])

        return thisTree

File: static/bleextop/application/libraries/applicationbi.php
<?php


require_once('static/bleextop/application/core/BleextBI'.EXT);

class ApplicationBI extends BleextBI{
	
	function ApplicationBI(){
		parent::BleextBI();
		
		$this->instance->load->model("applicationdao");
		$this->appdao = $this->instance->applicationdao;
	}
	
	public function getApplications($params){
		
		$apps = $this->appdao->getApplications($params);
		
		$tree = $this->buildTree($apps,"menu");
		
		return $tree->getRoot();

	}
	
	public function getTree(){
		
		$apps = $this->appdao->getAll();
		
		$tree = $this->buildTree($apps,"children");
		
		return $tree->getRoot();
	}
	
	public function save($form){
		$form->date_updated = date("Y-m-d h:i:s");
		$form->date_created = date("Y-m-d h:i:s");
		$id = $this->appdao->save($form);
		
		if($id === false){
			return array("success"=>false,"message"=>"There was an error saving this application.");
		}else{
			$this->appdao->addPermissions($id);
			
			return array("success"=>true,"application_k"=>$id,"message"=>"Application successfully saved");
		}
	}
	
	public function update($form){
		$form->date_updated = date("Y-m-d h:i:s");
		$success = $this->appdao->update($form);
		
		if($success){
			return array("success"=>true,"application_k"=>$form->application_k,"message"=>"Application successfully updated");
		}else{
			return array("success"=>false,"message"=>"There was an error updating this application.");
		}
	}
	
	public function remove($application_k){
		$this->appdao->remove($application_k);
	}
	
	public function move($form){
		$success = $this->appdao->move($form);
		if($success){
			return array("success"=>true,"message"=>"Application successfully moved");
		}else{
			return array("success"=>false,"message"=>"There was an error moving this application.");
		}
	}
	
	private function buildTree($apps,$text){
		$CI =& get_instance();
		$CI->load->library("tree");
		
		$temp = array();
		foreach($apps as $app){
			$iconCls = "";
			if($app["configurations"]){
				$conf = json_decode($app["configurations"]);
				if($conf){
					if(property_exists($conf,"iconCls")){
						$iconCls = $conf->iconCls;
					}
				}
			}
			array_push($temp,array(
				"text"			=> $app["name"],
				"name"			=> $app["name"],
				"application_k"	=> $app["application_k"],
				"application_parent_k"=> $app["application_parent_k"],
				"klass"			=> $app["klass"],
				"description"	=> $app["description"],
				"configurations"=> $app["configurations"],
				"active"		=> $app["active"],
				"iconCls"		=> $iconCls
			));
		}
		
		// Creating the Tree
		$tree = new Tree();
		$tree->setChildProperty($text);
		$tree->setIdProperty("application_k");
		foreach($temp as $app){
			$tree->addChild($app,$app["application_parent_k"]);
		}
		return $tree;
	}
}